hphp/runtime/vm/jit/vasm-arm.cpp

```diff
@@ -488,91 +480,72 @@ void Vgen::emit(const call& i) {
   }
 }
 
-void Vgen::emit(const callm& i) {
-  a->Ldr(rAsm, M(i.target));
-  a->Blr(rAsm);
-}
-
 void Vgen::emit(const calls& i) {
   emitSmashableCall(a->code(), env.meta, i.target);
 }
 
 ///////////////////////////////////////////////////////////////////////////////
 
-void Vgen::emit(const stublogue& i) {
-  // Push FP, LR always regardless of i.saveframe (makes SP 16B aligned)
-  emit(pushp{rfp(), rlr()});
-}
-
-void Vgen::emit(const stubret& i) {
-  if (i.saveframe) {
-    emit(popp{rfp(), rlr()});
-  } else {
-    emit(popp{PhysReg(rAsm), rlr()});
-  }
-  emit(ret{});
-}
-
-void Vgen::emit(const callstub& i) {
-  emit(call{i.target, i.args});
-}
-
 void Vgen::emit(const callfaststub& i) {
   emit(call{i.target, i.args});
   emit(syncpoint{i.fix});
 }
 
-void Vgen::emit(const tailcallstub& i) {
-  // Pop off FP/LR pair and jump to target
-  emit(popp{rfp(), rlr()});
-  emit(jmpi{i.target, i.args});
+void Vgen::emit(const callstub& i) {
+  emit(call{i.target, i.args});
 }
 
 ///////////////////////////////////////////////////////////////////////////////
 
-void Vgen::emit(const phplogue& i) {
-  // Save LR in m_savedRip on the current VM frame pointed by 'i.fp'
-  a->Str(X(rlr()), X(i.fp)[AROFF(m_savedRip)]);
-}
-
-void Vgen::emit(const phpret& i) {
-  a->Ldr(X(rlr()), X(i.fp)[AROFF(m_savedRip)]);
-  if (!i.noframe) {
-    a->Ldr(X(i.d), X(i.fp)[AROFF(m_sfp)]);
-  }
-  emit(ret{});
-}
-
 void Vgen::emit(const callphp& i) {
   emitSmashableCall(a->code(), env.meta, i.stub);
   emit(unwind{{i.targets[0], i.targets[1]}});
 }
 
-void Vgen::emit(const tailcallphp& i) {
-  // To make callee's return as caller's return, load the return address at
-  // i.fp[AROFF(m_savedRip)] into LR and jmp to target
-  a->Ldr(X(rlr()), X(i.fp)[AROFF(m_savedRip)]);
-  emit(jmpr{i.target, i.args});
-}
-
 void Vgen::emit(const callarray& i) {
   emit(call{i.target, i.args});
 }
 
 void Vgen::emit(const contenter& i) {
-  vixl::Label End;
+  vixl::Label stub, end;
+
+  // Jump past the stub below.
+  a->B(&end);
 
-  a->Adr(rAsm, &End);
-  a->Str(rAsm, X(i.fp)[AROFF(m_savedRip)]);
-  a->Blr(X(i.target));
-  a->bind(&End);
-  // m_savedRip will point here.
+  // We call into this stub from the end below. Take that LR and store it in
+  // m_savedRip. Then jump to the target.
+  a->bind(&stub);
+  a->Str(X(rlr()), M(i.fp[AROFF(m_savedRip)]));
+  a->Br(X(i.target));
+
+  // Call to stub above and then unwind.
+  a->bind(&end);
+  a->Bl(&stub);
   emit(unwind{{i.targets[0], i.targets[1]}});
 }
 
+///////////////////////////////////////////////////////////////////////////////
+
+void Vgen::emit(const calltc& i) {
+  vixl::Label stub;
+
+  // Preserve the FP and LR (exittc) on the stack.
```

Comment:
I think it'd be better to push and pop `rAsm` instead, then.  We should only ever call `leavetc{}` from the top frame in a VM reentry, which means that `rvmfp()` should be the same as it was at the `calltc{}` anyway.  Doing this save/restore could hide incorrect uses of `leavetc{}` (and also I think it's just less ambiguous if we avoid doing unnecessary work, for later code archaeology purposes).